lib/client.js

```diff
@@ -440,70 +437,62 @@ Client.prototype.eachRow = function () {
  * @param {function} [callback], executes callback(err) after all rows have been received or if there is an error
  * @returns {types.ResultStream}
  */
-Client.prototype.stream = function () {
-  var args = Array.prototype.slice.call(arguments);
-  if (typeof args[args.length-1] !== 'function') {
-    //the callback is not required
-    args.push(utils.noop);
-  }
-  args = utils.parseCommonArgs.apply(null, args);
+Client.prototype.stream = function (query, params, options, callback) {
+  callback = utils.bindDomain(callback || utils.noop);
```

Comment:
In this case, `callback` is an optional parameter (normally shouldn't be used) as you have `stream.on('end', handler)` event.
Optional parameters are left-to-right, so `callback` will only be defined if `params` and `options` are defined.

The only method that breaks the left-to-right rule is `eachRow()`, because the last `callback` is optional but `rowCallback` isn't (I don't like it but its legacy and we have to support it, at least until the next major version). In that case, I've included the overloads in the doc.